### ISSUE TITLE
Runtime error: Index out of bounds (IndexError)

### ISSUE THREAD

**Issue by zachad** [crystal-lang/crystal-mysql#3]
I'm getting an IndexError while trying to connect to my local MySQL 5.1.73 server. 

```
Index out of bounds (IndexError)
[4457383] *CallStack::unwind:Array(Pointer(Void)) +87
[4457274] *CallStack#initialize:Array(Pointer(Void)) +10
[4457226] *CallStack::new:CallStack +42
[4427528] *raise<IndexError>:NoReturn +24
[4478993] *Slice(UInt8) +161
[4585921] *MySql::Protocol::HandshakeV10::read<MySql::ReadPacket>:MySql::Protocol::HandshakeV10 +817
[4571449] *MySql::Connection#read_packet<MySql::Protocol::HandshakeV10:Class>:MySql::Protocol::HandshakeV10 +41
[4570542] *MySql::Connection#initialize<DB::Database>:UInt8 +686
[4569827] *MySql::Connection::new<DB::Database>:MySql::Connection +131
[4569693] *MySql::Driver#build_connection<DB::Database>:MySql::Connection +13
[4569665] *DB::Database#initialize<DB::Driver+, URI>:MySql::Connection +33
[4569603] *DB::Database::new<DB::Driver+, URI>:DB::Database +115
[4536132] *DB::build_database<URI>:DB::Database +36
[4536094] *DB::build_database<String>:DB::Database +14
[4536070] *DB::open<String>:DB::Database +6
[4402599] ???
[4437977] main +41
[140094784847133] __libc_start_main +253
[4398601] ???
```

I've traced the error back to this line:

<https://github.com/crystal-lang/crystal-mysql/blob/master/src/mysql/packets.cr#L27>

When the variable `auth_plugin_data_length` is set to 0, subtracting 8 from it results in the value 248 and not -8. I think because it is unsigned and wraps around?  Should `auth_plugin_data_length` be a signed integer?

The evaluation of `{13, auth_plugin_data_length - 8}.max` looks at `{13, 248}` and returns 248, rather than 13.  And we try to read `auth_data[8,247]` which throws the IndexError.

Hardcoding that line to `packet.read(auth_data[8,12])` avoids the crash and everything looks good after that.

I could submit a pull request to just hard-code that `[8,12]`, but I'm sure it's in there for a good reason.  Thanks for taking a look!

**Comment by bcardiff** [comment]
changing 
`{13, auth_plugin_data_length - 8}.max - 1]`
to
`auth_plugin_data_length < 22 ? 12 : (auth_plugin_data_length - 9)`
should be sign-safe.

or changing `auth_plugin_data_length` to `auth_plugin_data_length.to_i16` (or `to_i32`) should be enough since auth_plugin_data_length is `UInt8`.

There is no further reason than trying to implement the internals of mysql protocol. And, as you see there are chances to miss some details :-(. Thanks for jumping in crystal-mysql!

If the latest proposal work feel free to submit a PR. :-)

**Comment by zachad** [comment]
I've been having fun exploring Crystal more and more, and it's awesome to see all the work you have put in!  And I'm all for implementing internal protocols. No better way to learn what's happening under the hood. See PR #4 hopefully working around this issue.

**Comment by bcardiff** [comment]
PR merged at 6db306e.

I had fun with mysql internals too.

There are plenty of things pending to be implemented. And we try to move forward crystal-db family all together, but feel free to collaborate with the missing parts you come across. Some parts might be worth a discussion towards a general api though.

Thanks for the charming words 💚
